### ISSUE TITLE
Support reductions over arrays of atomics

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#12549]
As a Chapel programmer, I would like to be able to compute reductions over arrays of atomics because it is a useful thing to do and seems arbitrary that I cannot do it.  As a motivating example, maybe I have a per-locale atomic array and I want to compute the reduction over all locales of the atomic arrays.  Or I have an array of atomics that serve as buckets/counters and I want to sum them all up to get a total count.

Note that this feature request may relate to issue #8847.

**Associated Future Test(s):**
[`test/reductions/reduceAtomics.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/reductions/reduceAtomics.chpl
          ) #12547 

### Configuration Information

- Output of `chpl --version`: `chpl version 1.19.0 pre-release (2cec9f8)`

**Comment by bradcray** [comment]
Attn: @vasslitvinov, @ronawho

**Comment by ronawho** [comment]
You can do `+reduce A.read();` today:
```chpl 
var A: [1..10] atomic int;
A.write(10);
writeln(+reduce A.read()); // prints "100"
```
 but presumably you want to be able to do this without the promoted `read()`?

**Comment by ronawho** [comment]
Hmm, if we did support `+reduce A` without the `read()` what would you expect the result type to be?

**Comment by bradcray** [comment]
I don't think adding a `.read()` works for the array-of-arrays of atomics case...

I'd expect the result of reducing an array of `atomic t` to simply to be `t`.

**Comment by ronawho** [comment]
Yeah, good point. For the array-of-arrays case you'd have to do something like:

```chpl
var A: [1..2][1..2] atomic int;
[a in A] a.write(10);
writeln([a in A] +reduce a.read()); // prints "20 20"
```

But that's definetly more of a pain, and we've talked about wanting operator overloads anyways, so that's probably the right way to go.
